// @syncfusion/ej2-pivotview/src/common/conditionalformatting/conditional-formatting.js
import { Dialog } from '@syncfusion/ej2-popups';
import { isNullOrUndefined as isNaN, createElement, extend, remove, addClass } from '@syncfusion/ej2-base';
import { Button } from '@syncfusion/ej2-buttons';
import { ColorPicker } from '@syncfusion/ej2-inputs';
import { DropDownList } from '@syncfusion/ej2-dropdowns';
import * as cls from '../../common/base/css-constant';
/**
 * Module to render Conditional Formatting Dialog
 */
/** @hidden */
var ConditionalFormatting = /** @class */ (function () {
    /** Constructor for conditionalformatting module */
    function ConditionalFormatting(parent) {
        this.parent = parent;
        this.parent.conditionalFormattingModule = this;
        this.parentID = this.parent.element.id;
        this.dialog = null;
        this.fieldsDropDown = [];
        this.conditionsDropDown = [];
        this.fontNameDropDown = [];
        this.fontSizeDropDown = [];
        this.fontColor = [];
        this.backgroundColor = [];
        this.newFormat = [];
    }
    /**
     * To get module name.
     * @returns string
     */
    ConditionalFormatting.prototype.getModuleName = function () {
        return 'conditionalformatting';
    };
    ConditionalFormatting.prototype.createDialog = function () {
        var _this = this;
        if (document.querySelector('#' + this.parentID + 'conditionalformatting') !== null) {
            remove(document.querySelector('#' + this.parentID + 'conditionalformatting'));
        }
        this.parent.element.appendChild(createElement('div', {
            id: this.parentID + 'conditionalformatting',
        }));
        var buttonModel = [
            {
                'click': function () {
                    var format = {
                        conditions: 'LessThan',
                        value1: 0,
                        style: {
                            backgroundColor: 'white',
                            color: 'black',
                            fontFamily: 'Arial',
                            fontSize: '12px'
                        }
                    };
                    _this.refreshConditionValues();
                    _this.newFormat.push(format);
                    _this.addFormat();
                },
                buttonModel: {
                    cssClass: this.parent.isAdaptive ? (cls.FORMAT_ROUND_BUTTON + ' ' + cls.FORMAT_CONDITION_BUTTON) :
                        cls.FORMAT_CONDITION_BUTTON,
                    iconCss: cls.ICON + ' ' + cls.ADD_ICON_CLASS,
                    content: this.parent.isAdaptive ? '' : this.parent.localeObj.getConstant('condition'),
                }
            },
            {
                'click': function () {
                    _this.refreshConditionValues();
                    _this.parent.setProperties({ dataSource: { conditionalFormatSettings: _this.newFormat } }, true);
                    _this.parent.renderPivotGrid();
                    _this.destroy();
                },
                buttonModel: {
                    cssClass: cls.FLAT_CLASS + ' ' + cls.FORMAT_APPLY_BUTTON,
                    content: this.parent.localeObj.getConstant('apply')
                }
            },
            {
                'click': function () {
                    _this.destroy();
                    _this.newFormat = [];
                },
                buttonModel: {
                    cssClass: cls.FLAT_CLASS + ' ' + cls.FORMAT_CANCEL_BUTTON,
                    content: this.parent.localeObj.getConstant('cancel')
                }
            }
        ];
        if (this.parent.isAdaptive) {
            this.dialog = new Dialog({
                animationSettings: { effect: 'Zoom' }, isModal: true, width: '100%', height: '100%',
                showCloseIcon: false, closeOnEscape: false, enableRtl: this.parent.enableRtl,
                position: { X: 'center', Y: 'center' }, allowDragging: true, buttons: buttonModel,
                beforeOpen: function (args) {
                    _this.dialog.element.querySelector('.' + cls.DIALOG_HEADER).
                        setAttribute('title', _this.parent.localeObj.getConstant('conditionalFormating'));
                },
                cssClass: cls.FORMAT_DIALOG, header: this.parent.localeObj.getConstant('conditionalFormating'), target: document.body
            });
        }
        else {
            this.dialog = new Dialog({
                allowDragging: true, position: { X: 'center', Y: this.parent.element.offsetTop }, buttons: buttonModel,
                beforeOpen: function (args) {
                    _this.dialog.element.querySelector('.' + cls.DIALOG_HEADER).
                        setAttribute('title', _this.parent.localeObj.getConstant('conditionalFormating'));
                },
                cssClass: cls.FORMAT_DIALOG, isModal: false, closeOnEscape: true, enableRtl: this.parent.enableRtl,
                showCloseIcon: true, header: this.parent.localeObj.getConstant('conditionalFormating'), target: this.parent.element
            });
        }
        this.dialog.appendTo('#' + this.parentID + 'conditionalformatting');
    };
    ConditionalFormatting.prototype.refreshConditionValues = function () {
        for (var i = 0; i < this.newFormat.length; i++) {
            this.newFormat[i].value1 =
                Number(document.querySelector('#' + this.parentID + 'conditionvalue1' + i).value);
            this.newFormat[i].value2 =
                Number(document.querySelector('#' + this.parentID + 'conditionvalue2' + i).value);
        }
    };
    ConditionalFormatting.prototype.addFormat = function () {
        var format = createElement('div', { id: this.parentID + 'formatDiv', className: cls.FORMAT_NEW });
        for (var i = 0; i < this.newFormat.length; i++) {
            format.appendChild(this.createDialogElements(i));
        }
        this.dialog.setProperties({ 'content': format }, false);
        for (var i = 0; i < this.newFormat.length; i++) {
            this.renderDropDowns(i);
            this.renderColorPicker(i);
        }
    };
    ConditionalFormatting.prototype.createDialogElements = function (i) {
        var format = this.newFormat[i];
        var outerDiv = createElement('div', {
            id: this.parentID + 'outerDiv' + i, className: cls.FORMAT_OUTER
        });
        var button = createElement('button', {
            id: this.parentID + 'removeButton' + i, className: cls.FORMAT_DELETE_BUTTON,
            attrs: { 'title': this.parent.localeObj.getConstant('delete') }
        });
        outerDiv.appendChild(button);
        var innerDiv = createElement('div', { id: this.parentID + 'innerDiv', className: cls.FORMAT_INNER });
        var table = createElement('table', { id: this.parentID + 'cftable', className: cls.FORMAT_TABLE });
        var tRow = createElement('tr');
        var td = createElement('td');
        var valuelabel = createElement('span', {
            id: this.parentID + 'valuelabel' + i, className: cls.FORMAT_VALUE_LABEL,
            innerHTML: this.parent.localeObj.getConstant('value')
        });
        td.appendChild(valuelabel);
        tRow.appendChild(td);
        table.appendChild(tRow);
        tRow = createElement('tr');
        td = createElement('td');
        var measureDropdown = createElement('div', { id: this.parentID + 'measure' + i });
        var measureInput = createElement('input', {
            id: this.parentID + 'measureinput' + i,
            attrs: { 'type': 'text', 'tabindex': '1' }
        });
        measureDropdown.appendChild(measureInput);
        td.appendChild(measureDropdown);
        tRow.appendChild(td);
        td = createElement('td');
        var conditionDropdown = createElement('div', { id: this.parentID + 'condition' });
        var conditionInput = createElement('input', {
            id: this.parentID + 'conditioninput' + i,
            attrs: { 'type': 'text', 'tabindex': '1' }
        });
        conditionDropdown.appendChild(conditionInput);
        td.appendChild(conditionDropdown);
        tRow.appendChild(td);
        td = createElement('td');
        var style = !(format.conditions === 'Between' || format.conditions === 'NotBetween') ? 'display:none; width:10px' : '';
        var value1 = createElement('input', {
            id: this.parentID + 'conditionvalue1' + i,
            attrs: { 'type': 'text', 'tabindex': '1', 'value': !isNaN(format.value1) ? format.value1.toString() : '0' },
            styles: this.parent.isAdaptive ? style === '' ? 'width: 28%' : 'width: 100%' : style === '' ? 'width: 45px' : 'width: 120px',
            className: cls.INPUT + ' ' + cls.FORMAT_VALUE1
        });
        td.appendChild(value1);
        var valuespan = createElement('span', {
            id: this.parentID + 'valuespan' + i, className: cls.FORMAT_VALUE_SPAN,
            innerHTML: '&', styles: style
        });
        td.appendChild(valuespan);
        var value2 = createElement('input', {
            id: this.parentID + 'conditionvalue2' + i,
            attrs: { 'type': 'text', 'tabindex': '1', 'value': !isNaN(format.value2) ? format.value2.toString() : '0' },
            styles: (this.parent.isAdaptive && style === '') ? 'width: 28%' : style === '' ? 'width: 45px' : style,
            className: cls.INPUT + ' ' + cls.FORMAT_VALUE2
        });
        td.appendChild(value2);
        tRow.appendChild(td);
        table.appendChild(tRow);
        if (this.parent.isAdaptive) {
            innerDiv.appendChild(table);
            table = createElement('table', { id: this.parentID + 'cftable', className: cls.FORMAT_TABLE });
        }
        tRow = createElement('tr');
        td = createElement('td');
        var formatlabel = createElement('span', {
            id: this.parentID + 'formatlabel' + i, className: cls.FORMAT_LABEL, innerHTML: this.parent.localeObj.getConstant('formatLabel')
        });
        td.appendChild(formatlabel);
        tRow.appendChild(td);
        table.appendChild(tRow);
        tRow = createElement('tr');
        td = createElement('td');
        var fontNameDropdown = createElement('div', { id: this.parentID + 'fontname' });
        var fontNameInput = createElement('input', {
            id: this.parentID + 'fontnameinput' + i, attrs: { 'type': 'text', 'tabindex': '1' }
        });
        fontNameDropdown.appendChild(fontNameInput);
        td.appendChild(fontNameDropdown);
        tRow.appendChild(td);
        td = createElement('td');
        var fontSizeDropdown = createElement('div', { id: this.parentID + 'fontsize' });
        var fontSizeInput = createElement('input', {
            id: this.parentID + 'fontsizeinput' + i, attrs: { 'type': 'text', 'tabindex': '1' }
        });
        fontSizeDropdown.appendChild(fontSizeInput);
        td.appendChild(fontSizeDropdown);
        tRow.appendChild(td);
        if (this.parent.isAdaptive) {
            table.appendChild(tRow);
            tRow = createElement('tr');
            table.appendChild(tRow);
            tRow = createElement('tr');
        }
        td = createElement('td');
        var colorPicker1 = createElement('input', {
            id: this.parentID + 'fontcolor' + i, attrs: { 'type': 'color', 'tabindex': '1' }, className: cls.FORMAT_FONT_COLOR
        });
        td.appendChild(colorPicker1);
        var colorPicker2 = createElement('input', {
            id: this.parentID + 'backgroundcolor' + i, attrs: { 'type': 'color', 'tabindex': '1' }, className: cls.FORMAT_BACK_COLOR
        });
        td.appendChild(colorPicker2);
        tRow.appendChild(td);
        td = createElement('td');
        var valuePreview = createElement('div', {
            id: this.parentID + 'valuepreview' + i, className: cls.INPUT + ' ' + cls.FORMAT_VALUE_PREVIEW,
            innerHTML: '123.45',
        });
        td.appendChild(valuePreview);
        tRow.appendChild(td);
        table.appendChild(tRow);
        innerDiv.appendChild(table);
        outerDiv.appendChild(innerDiv);
        return outerDiv;
    };
    ConditionalFormatting.prototype.renderDropDowns = function (i) {
        var _this = this;
        var format = this.newFormat[i];
        var fields = [];
        fields.push({
            index: 0, name: this.parent.localeObj.getConstant('AllValues'),
            field: this.parent.localeObj.getConstant('AllValues')
        });
        for (var i_1 = 0; i_1 < this.parent.dataSource.values.length; i_1++) {
            fields.push({
                index: i_1 + 1,
                name: this.parent.dataSource.values[i_1].caption || this.parent.dataSource.values[i_1].name,
                field: this.parent.dataSource.values[i_1].name
            });
        }
        var value = isNaN(format.measure) ? this.parent.localeObj.getConstant('AllValues') : format.measure;
        this.fieldsDropDown[i] = new DropDownList({
            dataSource: fields, fields: { text: 'name' },
            value: value, width: this.parent.isAdaptive ? '100%' : '120px',
            popupHeight: '200px',
            change: function (args) {
                _this.newFormat[i].measure = args.value.toString() === _this.parent.localeObj.getConstant('AllValues') ?
                    undefined : args.value.toString();
            }
        });
        this.fieldsDropDown[i].appendTo('#' + this.parentID + 'measureinput' + i);
        var conditions = [
            { value: 'LessThan', name: this.parent.localeObj.getConstant('LessThan') },
            { value: 'LessThanOrEqualTo', name: this.parent.localeObj.getConstant('LessThanOrEqualTo') },
            { value: 'GreaterThan', name: this.parent.localeObj.getConstant('GreaterThan') },
            { value: 'GreaterThanOrEqualTo', name: this.parent.localeObj.getConstant('GreaterThanOrEqualTo') },
            { value: 'Equals', name: this.parent.localeObj.getConstant('Equals') },
            { value: 'NotEquals', name: this.parent.localeObj.getConstant('NotEquals') },
            { value: 'Between', name: this.parent.localeObj.getConstant('Between') },
            { value: 'NotBetween', name: this.parent.localeObj.getConstant('NotBetween') }
        ];
        value = isNaN(format.conditions) ? 'LessThan' : format.conditions;
        this.conditionsDropDown[i] = new DropDownList({
            dataSource: conditions, fields: { value: 'value', text: 'name' },
            value: value, width: this.parent.isAdaptive ? '100%' : '120px',
            popupHeight: '200px', popupWidth: '200px',
            change: function (args) {
                _this.newFormat[i].conditions = args.value;
                if (args.value === 'Between' || args.value === 'NotBetween') {
                    document.querySelector('#' + _this.parentID + 'valuespan' + i).style.display = 'inline-block';
                    document.querySelector('#' + _this.parentID + 'valuespan' + i).style.width =
                        _this.parent.isAdaptive ? '10%' : '10px';
                    document.querySelector('#' + _this.parentID + 'conditionvalue2' + i).style.display = 'inline-block';
                    document.querySelector('#' + _this.parentID + 'conditionvalue2' + i).style.width =
                        _this.parent.isAdaptive ? '28%' : '45px';
                    document.querySelector('#' + _this.parentID + 'conditionvalue1' + i).style.width =
                        _this.parent.isAdaptive ? '28%' : '45px';
                }
                else {
                    document.querySelector('#' + _this.parentID + 'valuespan' + i).style.display = 'none';
                    document.querySelector('#' + _this.parentID + 'conditionvalue2' + i).style.display = 'none';
                    document.querySelector('#' + _this.parentID + 'conditionvalue1' + i).style.width =
                        _this.parent.isAdaptive ? '100%' : '120px';
                }
            }
        });
        this.conditionsDropDown[i].appendTo('#' + this.parentID + 'conditioninput' + i);
        var fontNames = [
            { index: 0, name: 'Arial' }, { index: 1, name: 'San Serif' }, { index: 2, name: 'Impact' },
            { index: 3, name: 'Trebuchet MS' }, { index: 4, name: 'Serif' }, { index: 5, name: 'Verdana' },
            { index: 6, name: 'Courier New' }, { index: 7, name: 'Times New Roman' }, { index: 8, name: 'Tahoma' },
            { index: 9, name: 'Gerogia' }
        ];
        value = isNaN(format.style.fontFamily) ? 'Arial' : format.style.fontFamily;
        this.fontNameDropDown[i] = new DropDownList({
            dataSource: fontNames, fields: { text: 'name' },
            value: value, width: this.parent.isAdaptive ? '100%' : '120px',
            popupWidth: '150px', popupHeight: '200px',
            change: function (args) {
                _this.newFormat[i].style.fontFamily = args.value.toString();
                document.querySelector('#' + _this.parentID + 'valuepreview' + i).style.fontFamily = args.value;
            }
        });
        this.fontNameDropDown[i].appendTo('#' + this.parentID + 'fontnameinput' + i);
        var fontSize = [
            { index: 0, name: '9px' }, { index: 1, name: '10px' }, { index: 2, name: '11px' }, { index: 3, name: '12px' },
            { index: 4, name: '13px' }, { index: 5, name: '14px' }, { index: 6, name: '15px' }, { index: 6, name: '16px' }
        ];
        value = isNaN(format.style.fontSize) ? '12px' : format.style.fontSize;
        this.fontSizeDropDown[i] = new DropDownList({
            dataSource: fontSize, fields: { text: 'name' }, popupHeight: '200px',
            value: value, width: this.parent.isAdaptive ? '100%' : '120px',
            change: function (args) {
                _this.newFormat[i].style.fontSize = args.value.toString();
                document.querySelector('#' + _this.parentID + 'valuepreview' + i).style.fontSize = args.value;
            }
        });
        this.fontSizeDropDown[i].appendTo('#' + this.parentID + 'fontsizeinput' + i);
    };
    ConditionalFormatting.prototype.renderColorPicker = function (i) {
        var _this = this;
        var format = this.newFormat[i];
        var value = isNaN(format.style.color) ? 'black' : format.style.color;
        var color = this.isHex(value.substr(1)) ? value : this.colourNameToHex(value);
        document.querySelector('#' + this.parentID + 'valuepreview' + i).style.color = color;
        this.fontColor[i] = new ColorPicker({
            cssClass: cls.FORMAT_COLOR_PICKER, value: color, mode: 'Palette',
            change: function (args) {
                _this.newFormat[i].style.color = args.currentValue.hex;
                document.querySelector('#' + _this.parentID + 'valuepreview' + i).style.color =
                    args.currentValue.hex;
            }
        });
        this.fontColor[i].appendTo('#' + this.parentID + 'fontcolor' + i);
        addClass([this.fontColor[i].element.nextElementSibling.querySelector('.' + cls.SELECTED_COLOR)], cls.ICON);
        value = isNaN(format.style.backgroundColor) ? 'white' : format.style.backgroundColor;
        color = this.isHex(value.substr(1)) ? value : this.colourNameToHex(value);
        document.querySelector('#' + this.parentID + 'valuepreview' + i).style.backgroundColor = color;
        document.querySelector('#' + this.parentID + 'valuepreview' + i).style.fontFamily = format.style.fontFamily;
        document.querySelector('#' + this.parentID + 'valuepreview' + i).style.fontSize = format.style.fontSize;
        this.backgroundColor[i] = new ColorPicker({
            cssClass: cls.FORMAT_COLOR_PICKER, value: color, mode: 'Palette',
            change: function (args) {
                _this.newFormat[i].style.backgroundColor = args.currentValue.hex;
                document.querySelector('#' + _this.parentID + 'valuepreview' + i).style.backgroundColor =
                    args.currentValue.hex;
            }
        });
        this.backgroundColor[i].appendTo('#' + this.parentID + 'backgroundcolor' + i);
        addClass([this.backgroundColor[i].element.nextElementSibling.querySelector('.e-selected-color')], cls.ICON);
        var toggleBtn = new Button({
            iconCss: cls.ICON + ' ' + cls.FORMAT_DELETE_ICON,
            cssClass: cls.FLAT
        });
        toggleBtn.appendTo('#' + this.parentID + 'removeButton' + i);
        toggleBtn.element.onclick = function () {
            _this.newFormat.splice(i, 1);
            _this.addFormat();
        };
    };
    ConditionalFormatting.prototype.isHex = function (h) {
        var a = parseInt(h, 16);
        while (h.charAt(0) === '0') {
            h = h.substr(1);
        }
        return (a.toString(16) === h.toLowerCase() || (a === 0 && h === ''));
    };
    ConditionalFormatting.prototype.hexToRgb = function (hex) {
        var result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
        return result ? {
            r: parseInt(result[1], 16),
            g: parseInt(result[2], 16),
            b: parseInt(result[3], 16)
        } : null;
    };
    ConditionalFormatting.prototype.colourNameToHex = function (colour) {
        var colours = {
            'aliceblue': '#f0f8ff', 'antiquewhite': '#faebd7', 'aqua': '#00ffff', 'aquamarine': '#7fffd4',
            'azure': '#f0ffff', 'beige': '#f5f5dc', 'bisque': '#ffe4c4', 'black': '#000000',
            'blanchedalmond': '#ffebcd', 'blue': '#0000ff',
            'blueviolet': '#8a2be2', 'brown': '#a52a2a', 'burlywood': '#deb887', 'cadetblue': '#5f9ea0',
            'chartreuse': '#7fff00', 'chocolate': '#d2691e',
            'coral': '#ff7f50', 'cornflowerblue': '#6495ed', 'cornsilk': '#fff8dc', 'crimson': '#dc143c', 'cyan': '#00ffff',
            'darkblue': '#00008b', 'darkcyan': '#008b8b', 'darkgoldenrod': '#b8860b', 'darkgray': '#a9a9a9', 'darkgreen': '#006400',
            'darkkhaki': '#bdb76b', 'darkmagenta': '#8b008b', 'darkolivegreen': '#556b2f',
            'darkorange': '#ff8c00', 'darkorchid': '#9932cc', 'darkred': '#8b0000', 'darksalmon': '#e9967a', 'darkseagreen': '#8fbc8f',
            'darkslateblue': '#483d8b', 'darkslategray': '#2f4f4f', 'darkturquoise': '#00ced1',
            'darkviolet': '#9400d3', 'deeppink': '#ff1493', 'deepskyblue': '#00bfff', 'dimgray': '#696969', 'dodgerblue': '#1e90ff',
            'firebrick': '#b22222', 'floralwhite': '#fffaf0', 'forestgreen': '#228b22', 'fuchsia': '#ff00ff',
            'gainsboro': '#dcdcdc', 'ghostwhite': '#f8f8ff', 'gold': '#ffd700', 'goldenrod': '#daa520',
            'gray': '#808080', 'green': '#008000',
            'greenyellow': '#adff2f', 'honeydew': '#f0fff0', 'hotpink': '#ff69b4', 'indianred ': '#cd5c5c',
            'indigo': '#4b0082', 'ivory': '#fffff0',
            'khaki': '#f0e68c', 'lavender': '#e6e6fa', 'lavenderblush': '#fff0f5', 'lawngreen': '#7cfc00', 'lemonchiffon': '#fffacd',
            'lightblue': '#add8e6', 'lightcoral': '#f08080', 'lightcyan': '#e0ffff', 'lightgoldenrodyellow': '#fafad2',
            'lightgrey': '#d3d3d3', 'lightgreen': '#90ee90', 'lightpink': '#ffb6c1', 'lightsalmon': '#ffa07a', 'lightseagreen': '#20b2aa',
            'lightskyblue': '#87cefa', 'lightslategray': '#778899', 'lightsteelblue': '#b0c4de',
            'lightyellow': '#ffffe0', 'lime': '#00ff00', 'limegreen': '#32cd32', 'linen': '#faf0e6',
            'magenta': '#ff00ff', 'maroon': '#800000', 'mediumaquamarine': '#66cdaa', 'mediumblue': '#0000cd', 'mediumorchid': '#ba55d3',
            'mediumpurple': '#9370d8', 'mediumseagreen': '#3cb371', 'mediumslateblue': '#7b68ee',
            'mediumspringgreen': '#00fa9a', 'mediumturquoise': '#48d1cc', 'mediumvioletred': '#c71585', 'midnightblue': '#191970',
            'mintcream': '#f5fffa', 'mistyrose': '#ffe4e1', 'moccasin': '#ffe4b5', 'navajowhite': '#ffdead', 'navy': '#000080',
            'oldlace': '#fdf5e6', 'olive': '#808000', 'olivedrab': '#6b8e23', 'orange': '#ffa500', 'orangered': '#ff4500',
            'orchid': '#da70d6',
            'palegoldenrod': '#eee8aa', 'palegreen': '#98fb98', 'paleturquoise': '#afeeee', 'palevioletred': '#d87093',
            'papayawhip': '#ffefd5',
            'peachpuff': '#ffdab9', 'peru': '#cd853f', 'pink': '#ffc0cb', 'plum': '#dda0dd', 'powderblue': '#b0e0e6', 'purple': '#800080',
            'rebeccapurple': '#663399', 'red': '#ff0000', 'rosybrown': '#bc8f8f', 'royalblue': '#4169e1',
            'saddlebrown': '#8b4513', 'salmon': '#fa8072', 'sandybrown': '#f4a460', 'seagreen': '#2e8b57',
            'seashell': '#fff5ee', 'sienna': '#a0522d',
            'silver': '#c0c0c0', 'skyblue': '#87ceeb', 'slateblue': '#6a5acd', 'slategray': '#708090', 'snow': '#fffafa',
            'springgreen': '#00ff7f',
            'steelblue': '#4682b4', 'tan': '#d2b48c', 'teal': '#008080', 'thistle': '#d8bfd8', 'tomato': '#ff6347', 'turquoise': '#40e0d0',
            'violet': '#ee82ee', 'wheat': '#f5deb3', 'white': '#ffffff', 'whitesmoke': '#f5f5f5', 'yellow': '#ffff00',
            'yellowgreen': '#9acd32'
        };
        if (typeof colours[colour.toLowerCase()] !== 'undefined') {
            return colours[colour.toLowerCase()];
        }
        return '#d5d5d5';
    };
    /**
     * To create Conditional Formatting dialog.
     * @returns void
     * @hidden
     */
    ConditionalFormatting.prototype.showConditionalFormattingDialog = function () {
        this.newFormat = [];
        for (var i = 0; i < this.parent.dataSource.conditionalFormatSettings.length; i++) {
            this.newFormat.push(extend({}, this.parent.dataSource.conditionalFormatSettings[i].properties, null, true));
        }
        this.createDialog();
        this.dialog.refresh();
        this.addFormat();
    };
    /**
     * To destroy the Conditional Formatting dialog
     * @returns void
     * @hidden
     */
    ConditionalFormatting.prototype.destroy = function () {
        if (this.dialog && !this.dialog.isDestroyed) {
            this.dialog.hide();
            for (var i = 0; i < this.newFormat.length; i++) {
                if (this.fontColor[i] && !this.fontColor[i].isDestroyed) {
                    this.fontColor[i].destroy();
                }
                if (this.backgroundColor[i] && !this.backgroundColor[i].isDestroyed) {
                    this.backgroundColor[i].destroy();
                }
            }
            this.dialog.destroy();
        }
        else {
            return;
        }
    };
    return ConditionalFormatting;
}());
export { ConditionalFormatting };
