// @syncfusion/ej2-documenteditor/src/document-editor/implementation/dialogs/bookmark-dialog.js
import { ListView } from '@syncfusion/ej2-lists';
import { Button } from '@syncfusion/ej2-buttons';
import { createElement, L10n, setCulture } from '@syncfusion/ej2-base';
/**
 * The Bookmark dialog is used to add, navigate or delete bookmarks.
 */
var BookmarkDialog = /** @class */ (function () {
    /**
     * @private
     */
    function BookmarkDialog(viewer) {
        var _this = this;
        this.addBookmark = function () {
            _this.owner.owner.editorModule.insertBookmark(_this.textBoxInput.value);
            _this.owner.dialog.hide();
        };
        /* tslint:disable:no-any */
        this.selectHandler = function (args) {
            _this.textBoxInput.value = args.text;
            /* tslint:disable:no-any */
            var value = document.getElementById('bookmark_text_box');
            value.setSelectionRange(0, args.text.length);
            value.focus();
        };
        this.gotoBookmark = function () {
            _this.owner.selection.selectBookmark(_this.textBoxInput.value);
        };
        this.deleteBookmark = function () {
            _this.owner.owner.editorModule.deleteBookmark(_this.textBoxInput.value);
            _this.show();
        };
        this.owner = viewer;
    }
    /**
     * @private
     */
    BookmarkDialog.prototype.getModuleName = function () {
        return 'BookmarkDialog';
    };
    /**
     * @private
     */
    BookmarkDialog.prototype.initBookmarkDialog = function (localValue, bookmarks) {
        var instance = this;
        var id = this.owner.owner.containerId + '_insert_bookmark';
        this.target = createElement('div', { id: id, className: 'e-de-bookmark' });
        var headerValue = localValue.getConstant('Bookmark name') + ':';
        var dlgFields = createElement('div', { innerHTML: headerValue, className: 'e-bookmark-dlgfields' });
        this.target.appendChild(dlgFields);
        var commonDiv = createElement('div', { className: 'e-bookmark-common' });
        this.target.appendChild(commonDiv);
        var searchDiv = createElement('div', { className: 'e-bookmark-list' });
        commonDiv.appendChild(searchDiv);
        var textBoxDiv = createElement('div', { className: 'e-bookmark-textboxdiv' });
        searchDiv.appendChild(textBoxDiv);
        // tslint:disable-next-line:max-line-length
        this.textBoxInput = createElement('input', { className: 'e-input e-bookmark-textbox-input', id: 'bookmark_text_box' });
        this.textBoxInput.setAttribute('type', 'text');
        textBoxDiv.appendChild(this.textBoxInput);
        var listviewDiv = createElement('div', { className: 'e-bookmark-listViewDiv', id: 'bookmark_listview' });
        searchDiv.appendChild(listviewDiv);
        var arts = this.owner.bookmarks.keys;
        this.listviewInstance = new ListView({
            dataSource: bookmarks,
            cssClass: 'e-bookmark-listview',
        });
        this.listviewInstance.appendTo(listviewDiv);
        this.listviewInstance.addEventListener('select', this.selectHandler);
        var buttonDiv = createElement('div', { className: 'e-bookmark-button' });
        commonDiv.appendChild(buttonDiv);
        var addbuttonDiv = createElement('div', { className: 'e-bookmark-addbutton' });
        buttonDiv.appendChild(addbuttonDiv);
        var addButtonElement = createElement('button', { innerHTML: 'Add', id: 'add' });
        addbuttonDiv.appendChild(addButtonElement);
        var addbutton = new Button({ cssClass: 'e-button-custom' });
        addbutton.appendTo(addButtonElement);
        addButtonElement.addEventListener('click', this.addBookmark);
        var deleteButtonDiv = createElement('div', { className: 'e-bookmark-deletebutton' });
        buttonDiv.appendChild(deleteButtonDiv);
        var deleteButtonElement = createElement('button', { innerHTML: 'Delete', id: 'delete' });
        deleteButtonDiv.appendChild(deleteButtonElement);
        var deletebutton = new Button({ cssClass: 'e-button-custom' });
        deletebutton.appendTo(deleteButtonElement);
        deleteButtonElement.addEventListener('click', this.deleteBookmark);
        var gotoButtonDiv = createElement('div', { className: 'e-bookmark-gotobutton' });
        buttonDiv.appendChild(gotoButtonDiv);
        var gotoButtonElement = createElement('button', { innerHTML: 'Go To', id: 'goto' });
        gotoButtonDiv.appendChild(gotoButtonElement);
        var gotobutton = new Button({ cssClass: 'e-button-custom' });
        gotobutton.appendTo(gotoButtonElement);
        gotoButtonElement.addEventListener('click', this.gotoBookmark);
    };
    /**
     * @private
     */
    BookmarkDialog.prototype.show = function () {
        var bookmarks = this.owner.getBookmarks();
        var localObj = new L10n('documenteditor', this.owner.owner.defaultLocale);
        localObj.setLocale(this.owner.owner.locale);
        setCulture(this.owner.owner.locale);
        // if (!this.target) {
        this.initBookmarkDialog(localObj, bookmarks);
        //}
        this.owner.dialog.header = 'Bookmark';
        this.owner.dialog.height = 'auto';
        this.owner.dialog.width = 'auto';
        this.owner.dialog.content = this.target;
        this.owner.dialog.beforeOpen = this.owner.updateFocus;
        this.owner.dialog.close = this.owner.updateFocus;
        this.owner.dialog.buttons = [{
                click: this.removeObjects.bind(this),
                buttonModel: { content: localObj.getConstant('Cancel'), cssClass: 'e-flat e-hyper-insert', isPrimary: true }
            }];
        this.owner.dialog.dataBind();
        this.owner.dialog.show();
    };
    BookmarkDialog.prototype.removeObjects = function () {
        this.owner.dialog.hide();
    };
    /**
     * @private
     */
    BookmarkDialog.prototype.destroy = function () {
        if (this.textBoxInput) {
            this.textBoxInput.remove();
            this.textBoxInput = undefined;
        }
        if (this.listviewInstance) {
            this.listviewInstance.destroy();
            this.listviewInstance = undefined;
        }
    };
    return BookmarkDialog;
}());
export { BookmarkDialog };
